### PR TITLE
Fix issue 1081

### LINKED ISSUE #1081: aliasing expressions scheduled to different loops

### PR DESCRIPTION
**Pull request by FabioLuporini**
fixes #1081 

basically a performance issue due to too much Clusters fusion -- Clusters with ConditionalDimensions shouldn't be fused with so called dense Clusters, as they won't be optimized by the DSE

This PR fixes this

Issue was introduced by #1026 . Was fixing a thing and I ended up causing this. Anyway, now suitable tests added so we should be golden

### CODE REVIEW

File: tests/test_dimension.py [hunk 1]
```diff
@@ -819,3 +819,79 @@ def test_implicit_dims(self):
...
+        h = Function(name='h', grid=grid)
+
+        ctime = ConditionalDimension(name='ctime', parent=time, condition=time > 4)
+
+        eqns = [Eq(f.forward, f + 1),
+                Eq(h, f + 1),
+                Eq(g, f + 1, implicit_dims=[ctime]),
+                Eq(f.forward, f + 1, implicit_dims=[ctime]),
+                Eq(f.forward, f + 1),
+                Eq(g, f + 1)]
```
Comment (mloubout) [inline tests/test_dimension.py]: Not sure I follow, there is duplicate equation on purpose?

Comment (FabioLuporini) [reply tests/test_dimension.py]: All it matters is that the RHSs are identical

Semantically this test is stupid.

Practically it's useful to see that guarded and un-guarded Clusters are
1. processed separately
2. optimized as expected

Comment (mloubout) [reply tests/test_dimension.py]: Oh ok, it's super confusing to read the equations :D

Comment (FabioLuporini) [reply tests/test_dimension.py]: Think of them as statements. 

I agree the example is stupid, but at least it's concise

**Pull request merged**